oonib/testhelpers/http_helpers.py

```diff
@@ -168,7 +241,336 @@ def all(self, length, keyword):
             length = 100000
         self.write(self.genRandomPage(length, keyword))
 
+
+def encodeResponse(response):
+    body = None
+    body_length = 0
+    if getattr(response, 'body', None):
+        body = response.body
+        body_length = len(response.body)
+    return {
+        'headers':
+            {k.lower(): v for k, v in response.headers.getAllRawHeaders()},
+        'code': response.code,
+        'body_length': body_length,
+        'body': body
+    }
+
+
+def encodeResponses(response):
+    responses = []
+    responses += [encodeResponse(response)]
+    if response.previousResponse:
+        responses += encodeResponses(response.previousResponse)
+    return responses
+
+
+TITLE_REGEXP = re.compile("<title>(.*?)</title>", re.IGNORECASE | re.DOTALL)
+
+def extractTitle(body):
+    m = TITLE_REGEXP.search(body, re.IGNORECASE | re.DOTALL)
+    if m:
+        return unicode(m.group(1), errors='ignore')
+    return ''
+
+
+class TCPConnectProtocol(Protocol):
+    def connectionMade(self):
+        self.transport.loseConnection()
+
+class TCPConnectFactory(Factory):
+    def buildProtocol(self, addr):
+        return TCPConnectProtocol()
+
+REQUEST_HEADERS = {
+    'User-Agent': ['Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, '
+                   'like Gecko) Chrome/47.0.2526.106 Safari/537.36'],
+    'Accept-Language': ['en-US;q=0.8,en;q=0.5'],
+    'Accept': ['text/html,application/xhtml+xml,application/xml;q=0.9,'
+               '*/*;q=0.8']
+}
+
+class TrueHeaders(Headers):
+    def __init__(self, rawHeaders=None):
+        self._rawHeaders = dict()
+        if rawHeaders is not None:
+            for name, values in rawHeaders.iteritems():
+                if type(values) is list:
+                    self.setRawHeaders(name, values[:])
+                elif type(values) is dict:
+                    self._rawHeaders[name.lower()] = values
+                elif type(values) is str:
+                    self.setRawHeaders(name, values)
+
+    def setRawHeaders(self, name, values):
+        if name.lower() not in self._rawHeaders:
+            self._rawHeaders[name.lower()] = dict()
+        self._rawHeaders[name.lower()]['name'] = name
+        self._rawHeaders[name.lower()]['values'] = values
+
+    def getAllRawHeaders(self):
+        for k, v in self._rawHeaders.iteritems():
+            yield v['name'], v['values']
+
+    def getRawHeaders(self, name, default=None):
```

Comment:
This is mapped directly to the interface of the `twisted.web.http_headers.Headers` class. It's in the plural because actually there can be multiple headers for a given header name.